modules/config/main.tf

```diff
@@ -0,0 +1,304 @@
+    resources = ["*"]
+
+    condition {
+      test     = "StringEquals"
+      variable = "ec2:ResourceTag/aws:autoscaling:groupName"
+      values   = [var.runner_autoscaling_group_name]
+    }
+  }
+
+  statement {
+    effect    = "Allow"
+    actions   = ["ssm:SendCommand"]
+    resources = [aws_ssm_document.reload_config.arn]
+  }
+}
+
+resource "aws_iam_role" "config_update_ssm_command" {
+  name               = "${var.name}-config-update"
+  assume_role_policy = data.aws_iam_policy_document.config_update_ssm_command_trust.json
+}
+
+resource "aws_iam_role_policy" "config_update_ssm_command" {
+  role   = aws_iam_role.config_update_ssm_command.name
+  name   = "ConfigUpdateSSMCommand"
+  policy = data.aws_iam_policy_document.config_update_ssm_command.json
+}
+
+resource "aws_ssm_document" "reload_config" {
+  name          = "${var.name}-reload-gitlab-configuration"
+  target_type   = "/AWS::EC2::Instance"
+  document_type = "Command"
+
+  content = jsonencode({
+    schemaVersion = "2.2"
+    description   = "Reload Gitlab Configuration."
+    parameters    = {},
+    mainSteps = [
+      {
+        action = "aws:runShellScript"
+        name   = "reloadGitlabConfiguration"
+        inputs = {
+          timeoutSeconds = "60"
+          runCommand = [
+            <<-EOF
+              echo "Validating Gitlab installation..."
+              hash jq gitlab-runner >/dev/null || {
+                echo "Runner is not ready yet, skipping update, as runner should pull configuration automatically on boot."
+                exit 0
+              }
+              echo "Pulling Gitlab token from SSM parameters..."
+              token=$(aws ssm get-parameters --names "${var.gitlab_token_ssm_key}" --with-decryption --region "${data.aws_region.current.name}" | jq -r ".Parameters | .[0] | .Value")
+
+              echo "Pulling file from S3 bucket..."
+              config_file="/etc/gitlab-runner/config.toml"
+              aws s3 cp "${local.config_uri}" "$${config_file}"
+
+              echo "Replacing tokens in configuration with stored Gitlab token..."
+              sed -i.bak s/__REPLACED_BY_USER_DATA__/`echo $token`/g "$${config_file}"
+
+              echo "Pulling extra files from S3 bucket (if any)..."
+              ${local.extra_files_sync_command}
+
+              ${var.post_reload_script}
+
+              echo "Restarting GitLab runner..."
+              gitlab-runner restart
+            EOF
+          ]
+        }
+      }
+    ]
+  })
+}
+
+resource "aws_cloudwatch_event_rule" "config_changes" {
+  name        = "${var.name}-config-changes"
+  description = "Watch configuration changes for Gitlab runner."
+
+  event_pattern = jsonencode({
+    source = ["aws.s3"]
+    detail = {
+      eventSource = ["s3.amazonaws.com"]
+      eventName   = ["PutObject", "CompleteMultipartUpload"]
+      requestParameters = {
+        bucketName = [aws_s3_bucket_object.config.bucket]
+        key = concat([aws_s3_bucket_object.config.key], [
```

Comment:
These keys are known up-front rather than being dynamic. This means that if we add a new extra file, there's a race condition between the cloudwatch rule being updated and the file in question being uploaded.

Perhaps the `aws_s3_bucket_object` should depend on this `aws_cloudwatch_event_rule`? I think we want to update the rule first and upload second: this allows blocks like

```hcl
post_reload_config = "(cd /extra-files/runners-config; zip -r ../runners-config.zip *)"
```

to actually work.